src/Recurr/Transformer/InstanceGenerator.php

```diff
@@ -0,0 +1,771 @@
+     */
+    protected $leapBug = false;
+
+    /**
+     * Constructor
+     */
+    public function __construct()
+    {
+        $this->leapBug = DateUtil::hasLeapYearBug();
+    }
+
+    /**
+     * Returns a generator that yields instances for the given rules and dates.
+     *
+     * @param DateTimeZone $timeZone The time zone to expand instances within.
+     * @param Rule[] $rrules A collection of iCal RRULE values to expand by.
+     * @param DateTime[] $rdates A collection of iCal RDATE values to include in the expansion.
+     * @param Rule[] $exrules A collection of iCal EXRULE values to expand and exclude.
+     * @param DateTime[] $exdates A collection of iCal EXDATE values to exclude in the expansion. (NOTE: EXDATE was
+     *   introduced in RFC2445 and deprecated in RFC5545. It is included here for legacy support because it is commonly
+     *   supported by popular cloud-based calendar systems like Google Calendar.
+     * @param bool $ignoreCounts Ignores the `COUNT` constraint of all rules given. This is useful if the `$seed`
+     *   represents a date in the middle of the series and using `COUNT` would be cause for an inaccurate end date.
+     * @param int|null $iterationLimit An iteration limit that, when provided, will serve as a fail-safe
+     *   that stops the generator when the limit is reached to prevent an infinitely large iterator.
+     * @return Generator A generator that yields {@link DateTime} values representing instance start time.
+     * @throws MissingData
+     */
+    public function generate(
+        DateTimeZone $timeZone,
+        array $rrules = [],
+        array $rdates = [],
+        array $exrules = [],
+        array $exdates = [],
+        $ignoreCounts = false,
+        $iterationLimit = null
+    ) {
+        if (empty($rrules) && empty($rdates) && empty($exrules) && empty($exdates)) {
+            return; // Empty generator.
+        }
+
+        // An array to hold all the generators for dates that we'll include.
+        $rGenerators = [];
+
+        // An array to hold all the generators for dates that we'll exclude.
+        $exGenerators = [];
+
+        // Add a generator for the RDATE collection
+        $rGenerators[] = $this->generateFromDates($rdates, $timeZone);
+
+        // Add a generator for each RRULE
+        foreach ($rrules as $rrule) {
+            $rGenerators[] = $this->generateFromRule(
+                $rrule,
+                $timeZone,
+                $ignoreCounts
+            );
+        }
+
+        // Add a generator for the EXDATE collection
+        $exGenerators[] = $this->generateFromDates($exdates, $timeZone);
+
+        // Add a generator for each EXRULE
+        foreach ($exrules as $exrule) {
+            $exGenerators[] = $this->generateFromRule(
+                $exrule,
+                $timeZone,
+                $ignoreCounts
+            );
+        }
+
+        $iterations = 0;
+
+        do {
+            /** @var Generator $generator */
```

Comment:
I feel like this is out of place. I usually see these **directly** above the iteration blocks/statements.